doc/gold.rst

```diff
@@ -0,0 +1,412 @@
+
+
+Language Elements
+=================
+
+A Gold-syntax unit consists of a series of top-level statements.
+
+Binding Statements
+------------------
+
+The following statements introduce bindings of names. They are all allowed at
+the top level of the file, and a subset are allowed inside executable code
+blocks.
+
+Top-level bindings are visible throughout the program. Block-level bindings are
+visible until the end of the scope in which they are introduced. Introducing a
+block-level binding with the same name as another binding, even in the same
+scope, will result in the previous binding being shadowed until the new binding
+goes out of scope, at which time the previous binding will be restored.
+
+It is an error to introduce a new top-level binding with the same name as
+another top-level binding.
+
+
+``use``
+~~~~~~~
+
+Top-level usage: ::
+
+  use <identifier>
+
+Locates a unit named with the given identifier, and makes its exported symbols
+available in the current unit in a namespace bound to the same name as the unit.
+If the unit cannot be located, a compilation error will be raised.
+
+The binding introduced by a ``use`` statement is not exported from the unit.
+
+Units have names derived from the name of the file they are defined in, and thus
+the allowed names for units may be further restricted by what characters are
+allowed in filenames on your system. For maximum portability, stick to
+alphanumeric characters and the character ``-`` (hyphen minus) in your unit
+names.
+
+
+``constant``
+~~~~~~~~~~~~
+
+Top-level / block-level usage: ::
+
+  constant <identifier> : <type> = <known-expression>
+
+Binds a name to a value known at compile time which does not allocate memory in
+the program image. The value will be inlined at usage sites. Top-level constant
+bindings are exported from the unit as symbols, and may be referenced in other
+units.
+
+The restriction to values which do not allocate memory means that arrays and
+strings cannot be declared as constant-bindings. It is possible to declare
+pointers and slices as constants through the use of certain functions exported
+from the built-in ``mem`` unit.
+
+
+``let``
+~~~~~~~
+
+Top-level usage ::
+
+  let [mut] <identifier> : <type> = <known-expression>
+
+Binds a name to a value known at compile time. Always allocates memory in the
+program image. Top-level let-bindings are exported from the unit as symbols
+which may be referenced in other units.
+
+Block-level usage: ::
+
+  let [mut] <identifier> : <type> = <expression>
+  let stash <identifier> : <type> = <known-expression>
+
+Binds a name to a value. In the first form, memory is allocated statically (i.e.
+memory is reserved, but the value is not included in the program image), and the
+value is computed and stored when the statement is executed. In the second form,
+memory is allocated in the program image with the initial value stored.
+
+By default, let-bindings are immutable, and may not be re-bound (though they may
```

Comment:
"and may not be re-bound": are you just rephrasing "immutable", or is there a nuanced difference between the two?

Reply:
This sentence is actually wrong! There is a difference, since "immutable" means "the memory location can't have its contents changed", while "can't be re-bound" means "the name can't be changed to point to a different memory location", but this is in direct contradiction to the section on block-level bindings you called out.

There should probably be a glossary of terms so that I'm not going Humpty-Dumpty and using words to mean what I mean them to mean.